types/wlr_screenshooter.c

```diff
@@ -41,27 +30,32 @@ static void handle_screenshot_resource_destroy(
 	}
 }
 
-static void output_frame_notify(struct wl_listener *listener, void *_data) {
+static void output_handle_frame(struct wl_listener *listener, void *_data) {
 	struct screenshot_state *state = wl_container_of(listener, state,
 		frame_listener);
-	struct wlr_renderer *renderer = state->screenshot->screenshooter->renderer;
 	struct wlr_output *output = state->screenshot->output;
+	struct wlr_renderer *renderer = wlr_backend_get_renderer(output->backend);
+	struct wl_shm_buffer *shm_buffer = state->shm_buffer;
 
-	wlr_output_make_current(output);
-	wlr_renderer_read_pixels(renderer, 0, 0, output->width, output->height,
-		state->pixels);
-
-	void *data = wl_shm_buffer_get_data(state->shm_buffer);
-	wl_shm_buffer_begin_access(state->shm_buffer);
-	copy_yflip(data, state->pixels + state->stride * (state->height - 1),
-		state->height, state->stride);
-	wl_shm_buffer_end_access(state->shm_buffer);
-
-	free(state->pixels);
-	wl_list_remove(&listener->link);
+	enum wl_shm_format format = wl_shm_buffer_get_format(shm_buffer);
+	int32_t width = wl_shm_buffer_get_width(shm_buffer);
+	int32_t height = wl_shm_buffer_get_height(shm_buffer);
+	int32_t stride = wl_shm_buffer_get_stride(shm_buffer);
+	void *data = wl_shm_buffer_get_data(shm_buffer);
+	wl_shm_buffer_begin_access(shm_buffer);
+	bool ok = wlr_renderer_read_pixels(renderer, format, stride, width, height,
+		0, 0, 0, 0, data);
+	wl_shm_buffer_end_access(shm_buffer);
+
+	if (!ok) {
+		wlr_log(L_ERROR, "Cannot read pixels");
+		goto cleanup;
```

Comment:
is it a problem that the `done` event never gets emitted in this case? The actual orbital screenshooter has a `fail` event for this case.

Reply:
It seems we have an older version of the protocol.